### ISSUE TITLE
Is it possible to get the coordinates of abstract, paragraph, headers, etc.?

### ISSUE THREAD

**Issue by qianhwan** [grobidOrg/grobid#456]
I'm wondering if there's a way to extract coordinates of other parts in a pdf (other than citation, tables, figures, etc.)? Thanks!

**Comment by kermitt2** [comment]
Hello @qianhwan !

Adding coordinates to more structures is not a big effort, a few lines to add (similar to the existing supported structures), to extend the web service parameters and to document... except for the structure in the header section which would need a bit more work because older - the header class is not using the `TaggingTokenClusteror` class yet, which is required to easily create bounding boxes for any sequence labelled by the sequence labeller. 

Which structure coordinates would you be interested in?

**Comment by lfoppiano** [comment]
For references. Some work have been done on the branch `desirtrack_coords`.  
https://github.com/kermitt2/grobid/tree/desirtrack_coords

The coordinates were implemented, plus the header and fulltext result was updated using the `TaggingTokenClusteror`. I cannot recall if it can be merged (perhaps requires some more work) but the coordinates were working as far as I remember. 

I've just align it with master ad I'm opening a PR, so that the review can be carried on easily.

**Comment by kermitt2** [comment]
Thanks @lfoppiano ! I will review the PR. 
But to clarify, there is nothing related to the fulltext, right?

**Comment by lfoppiano** [comment]
> But to clarify, there is nothing related to the fulltext, right?

Yes, sorry, just Header parser.

**Comment by lfoppiano** [comment]
Using this branch you can get the coordinates by 

POST http://localhost:8070/api/processFulltextDocument

and adding body as form-data: 
 - input: the input file
 - teiCoordinates: 

the parameter `teiCoordinates` requires one single value and can be repeated if required to get coordinates from more than one elements. 

The type of element are: 
persName (affiliations), formula (equations), figure (figures..), abstract, ref (references)

**Comment by lfoppiano** [comment]
For reference, to complete all the information (before they are lost forever), I'm also pointing to some code using this feature: https://github.com/DARIAH-ERIC/DESIR-CodeSprint-TrackA-TextMining/blob/master/src/main/java/org/dariah/desir/client/GrobidClient.java

**Comment by qianhwan** [comment]
Thank you!
> Which structure coordinates would you be interested in?

I'm thinking if there's a way to preserve all the coordinates info from `pdfalto`? Something like the `pdfalto` output xml and with segment tags.

**Comment by kermitt2** [comment]
Do you mean introducing an elements for each token with coordinates  in the final TEI (like have a tag `<w coords-"...">` for each word)? Or just coordinates to the structure elements already existing in the final TEI? 
Both are doable, because each token is keeping its original coordinates through all the process of GROBID, but the first one will make the TEI really unreadable (like ALTO files generated by pdfalto!).

**Comment by qianhwan** [comment]
> a tag `<w coords-"...">` for each word

Yeah I meant this, since it will allow more manipulation of the outputs. I agree the TEI would be totally unreadable though.
I had this question because I was trying to find the original location of some texts (sentences in paragraph for example) in a pdf, and `regex` matching doesn't seem to be the right tool for me since there could be some mismatches (symbols, citation number etc..).

**Comment by lfoppiano** [comment]
@qianhwan a quick response, I think the best solution here is to work directly with the `LayoutToken` objects and define a custom service or a submodule that output only the information you are looking for. It's not easy starting from scratch but it's doable. 

For example you can have a look at this example within the grobid project: https://github.com/kermitt2/grobid/blob/6699124232116fd210c604022a1f0256d951e31d/grobid-service/src/main/java/org/grobid/service/process/GrobidRestProcessFiles.java#L556 

or https://github.com/kermitt2/grobid-example

**Comment by qianhwan** [comment]
@lfoppiano Great! Thank you so much!

**Comment by robodasha** [comment]
Hi @qianhwan, did you manage to get this working? I need exactly the same information (coordinates for each word). I'd be happy to help/collaborate on this to get it done.

**Comment by qianhwan** [comment]
> Hi @qianhwan, did you manage to get this working? I need exactly the same information (coordinates for each word). I'd be happy to help/collaborate on this to get it done.

Hi @robodasha, I ended up adding a few lines in `TEIFormatter.java` so it also outputs the line coordinates for the parts I wanted, e.g:

```java
coord = teiElement("coords");
coord.appendChild(LayoutTokensUtil.getCoordsString(cluster.concatTokens));
head.appendChild(coord)
```

**Comment by ranihorev** [comment]
I'd love to get the coordinates of section titles (`<head>`). Is there a way to do that?
@qianhwan do you know by any chance?

**Comment by kermitt2** [comment]
@ranihorev since it is only 4 lines to add in `TEIFormatter.java` I've just made it with commit d5e5bc202db855299c7e5207f4ccbeebdb8afc3d 

```shell
curl --form input=@./NPJBCANCER-00349R.pdf --form teiCoordinates=head localhost:8070/api/processFulltextDocument
```

```xml
<div xmlns="http://www.tei-c.org/ns/1.0">
       <head coords="6,72.00,528.39,176.03,10.16">Medical management and treatment</head>
       <p>The patient was started on ursodiol 1200 mg/day for progressive pruritis on HD 12. 
....
```

Multi-line section titles will give several bounding boxes.

**Comment by ranihorev** [comment]
@kermitt2 thanks a lot!

**Comment by ranihorev** [comment]
I don't know if it's on purpose but the attribute value includes the text "coords='<the value>'". I can extract the actual value of course but I wanted to let you know :)

**Comment by kermitt2** [comment]
> I don't know if it's on purpose but the attribute value includes the text "coords=''". I can extract the actual value of course but I wanted to let you know :)

oops thank you @ranihorev sorry it's a bug, too fast :) 
It's fixed with ff48f01f13b1f4de2d11df522bdea784a29b292b

**Comment by ranihorev** [comment]
Just wanted to say thanks for the awesome tool!
We're now using it on Scihive.org to extract PDFs metadata and table of contents automatically :)
(it's open-source too btw)
